File: p7/p7.js
const fs = require('fs');
const { cursorTo } = require('readline');

const FILENAME = 'p7/p7_input.txt';

const getData = (fn) => fs.readFileSync(fn).toString().split('\n');

const initDir = { files: [] };
m = new Object({
  '/': new Object(initDir)
});

function accessM(l) {
  let localm = m;
  l.forEach((i) => {
    localm = localm[i];
  });
  return localm;
}

function formatFile(f) {
  const [filesize, filename] = f.split(' ');
  return [filename, filesize];
}

const makeFiletree = (arr) => {
  current_d = [];
  while (arr.length) {
    const com = arr.shift(0);
    if (com.slice(0, 1) === '$') {
      if (com.slice(2, 4) === 'cd') {
        if (com.slice(5) === '/') {
          current_d = ['/'];
          continue;
        }
        if (com.slice(5) === '..') {
          current_d = current_d.splice(0, current_d.length - 1);
        } else {
          current_d.push(com.slice(5));
        }
      }
    }
    if (com.slice(2, 4) === 'ls') {
      while (true) {
        if (!arr.length) {
          break;
        }
        if (arr[0][0] === '$') {
          break;
        } else {
          const outp = arr.shift();
          const targetDir = accessM(current_d);
          if (outp.slice(0, 3) === 'dir') {
            const listedDir = outp.slice(4);
            if (!(listedDir in targetDir)) {
              targetDir[listedDir] = { files: [] };
            }
          } else {
            const fileList = formatFile(outp);
            targetDir['files'].push(fileList);
          }
        }
      }
    }
  }
  return m;
};

function calculate_dir_size(name, m) {
  const sizeFiles = m['files'].reduce((x, y) => x + parseInt(y[1]), 0);
  const dirs = Object.keys(m)
    .filter((k) => k !== 'files')
    .map((k) => {
      return calculate_dir_size(k, m[k]);
    });
  const sizeDirs = dirs.reduce((x, y) => x + y[0], 0);
  const smallDirs = dirs.map((d) => d[1]);
  const size = sizeFiles + sizeDirs;
  if (size <= 100000) {
    smallDirs.push([name, size]);
  }
  return [size, smallDirs];
}

function gather_dir_sizes(name, m) {
  const sizeFiles = m['files'].reduce((x, y) => x + parseInt(y[1]), 0);
  const dirs = Object.keys(m)
    .filter((k) => k !== 'files')
    .map((k) => {
      return gather_dir_sizes(k, m[k]);
    });
  const sizeDirs = dirs.reduce((x, y) => x + y[0], 0);
  const dirNames = dirs.map((d) => d[1]);
  const size = sizeFiles + sizeDirs;
  dirNames.push([name, size]);
  return [size, dirNames];
}

function get_sum_small_dir_sizes(d) {
  if (!d.length) {
    return 0;
  }
  if (!Array.isArray(d[0])) {
    return d[1];
  }
  if (Array.isArray(d[0])) {
    return Object.keys(d).reduce(
      (x, k) => x + get_sum_small_dir_sizes(d[k]),
      0
    );
  } else {
  }
}

const process = (arr) => {
  const filetree = makeFiletree(arr);
  const [dirsize, smallFiles] = calculate_dir_size('/', filetree['/']);
  const output = get_sum_small_dir_sizes(smallFiles);
  return output;
};

const processP2 = (arr) => {
  const filetree = makeFiletree(arr);
  const [used_diskspace, size_by_dir] = gather_dir_sizes('/', filetree['/']);
  const total_diskspace = 70000000;
  const needed_free_diskspace = 30000000;
  const free_diskspace = total_diskspace - used_diskspace;
  const dirs_by_size = size_by_dir.flat(Infinity);
  const pairs = [];
  for (let i = 0; i < dirs_by_size.length; i += 2) {
    pairs.push(dirs_by_size.slice(i, i + 2));
  }
  elig_dir = pairs.filter((x) => free_diskspace + x[1] > needed_free_diskspace);
  const minDir = elig_dir.sort((x, y) => x[1] - y[1])[0];
  return minDir;
};

function main() {
  const data = getData(FILENAME);
  console.log(data);
  //const total = process(data);
  const totalP2 = processP2(data);
  console.log(totalP2);
}

main();

File: p7/p7.py
import csv
import copy
import pprint
pp = pprint.PrettyPrinter(indent=4)

FILENAME = "p7/p7_input.txt"

def get_data():
    output = []
    with open(FILENAME, 'r') as fd:
        reader = csv.reader(fd)
        for row in reader:
            output.append(next(iter(row),None))
    return output


init_dir = {"files":[]}
m = {
    "/": copy.deepcopy(init_dir)
}

def access_m(l):
    localm = m
    for i in l:
        localm = localm[i]
    return localm

def format_file(f):
    filesize, filename = f.split(" ")
    return tuple([filename,filesize])

def make_filetree(arr):
    current_d = []
    while arr:
        com = arr.pop(0)
        if com[0] == "$":
            if com[2:4] == "cd":
                if com[5:] == "/":
                    current_d = ["/"]
                    continue
                if com[5:] == "..":
                    current_d = current_d[:len(current_d)-1]
                else:
                    current_d.append(com[5:])
            if com[2:4] == "ls":
                while True:
                    if not arr:
                        break
                    if arr[0][0] == "$":
                        break
                    else:
                        outp = arr.pop(0)
                        if outp[:3] == "dir":
                            listed_dir = outp[4:]
                            if listed_dir not in access_m(current_d):
                                access_m(current_d)[listed_dir] = copy.deepcopy(init_dir)
                        else:
                            access_m(current_d)["files"].append(format_file(outp))                     
    return m

def calculate_dir_size(name, m):
    files = sum([int(i[1]) for i in m["files"]])
    dirs = [calculate_dir_size(k,m[k]) for k in m if k != "files"]
    dir_sizes = sum([int(i[0]) for i in dirs])
    small_dirs = [i[1] for i in dirs]
    size = files + dir_sizes
    if size <= 100000:
        small_dirs.append(tuple([name,size]))
    return [size,small_dirs]

def gather_dir_sizes(name, m):
    files = sum([int(i[1]) for i in m["files"]])
    dirs = [gather_dir_sizes(k,m[k]) for k in m if k != "files"]
    dir_sizes = sum([int(i[0]) for i in dirs])
    dirs_list = [i[1] for i in dirs]
    size = files + dir_sizes
    dirs_list.append(tuple([name,size]))
    return [size,dirs_list]

def get_sum_small_dir_sizes(d):
    if isinstance(d, tuple):
        return d[1]
    if isinstance(d, list):
        return sum([get_sum_small_dir_sizes(x) for x in d])
    else:
        pass

def make_list_dirs_by_size(d):
    output = []
    if isinstance(d, tuple):
        output.append(d)
    if isinstance(d, list):
        for x in d:
            output += make_list_dirs_by_size(x)
    else:
        pass
    return output

def process(arr):
    filetree = make_filetree(arr)
    dirsize, small_dirs = calculate_dir_size("/", filetree["/"])
    output = get_sum_small_dir_sizes(small_dirs)
    return output

def process_p2(arr):
    filetree = make_filetree(arr)
    used_diskspace, size_by_dir = gather_dir_sizes("/", filetree["/"])
    total_diskspace = 70000000
    needed_free_diskspace = 30000000
    free_diskspace = total_diskspace - used_diskspace
    dirs_by_size = make_list_dirs_by_size(size_by_dir)
    elig_dir = list(filter(lambda x: free_diskspace + x[1] > needed_free_diskspace, dirs_by_size))
    return sorted(elig_dir, key=lambda x: x[1])[0]

def main():
    arr = get_data()
    total = process(arr)
    total_p2 = process_p2(arr)
    print(total,total_p2)

main()